### DISCUSSION TITLE
Define interface implementations for any value or reference type

### DISCUSSION THREAD

**Discussion by br3nt** [dotnet/csharplang#7120] [Language Ideas]
This idea introduces two ideas:
* An interface can be implemented outside of a class/struct/record definition
* An interface implementation can be defined for any value or reference type.  E.g.:
  * Integral numeric types (`sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `nint`, `nuint`)
  * Floating point numeric types (`float`, `double`, `decimal`)
  * `bool`
  * `char`, and `string`
  * Enumeration types
  * Struct types
  * Tuple types
  * Nullable value types
  * Class types
  * Interface types
  * Delegate types
  * Record types
  * Arrays
  * `dynamic` types
  * Have I missed anything?

When an interface implementation is defined for a value or type, that value or type can be assigned or cast to that interface.

The implementation does not add to or extend the actual value or type with the methods/properties/etc defined for the interface. The implementation is only available when the value or type is assigned or cast to that interface.

The body of the implementation must implement all properties/methods/etc of the interface being implemented. 

An implementation cannot contain a constructor.

An implementation can't define additional properties/methods/etc that are not present on the interface. 

The implementation only has access to the public members of the value or type. This is because the value or type is not being redefined or extended and because the value or type may come from another assembly. 

The `this` keyword is used to access the value or type. Properties, methods, fields, ect can be accessed without the `this` keyword, except in cases where it matches a member on the interface.

**What does this solve?**

This functionality allows us to define how any value or object maps to an interface.  Because we don't need to own the value or type definition to define an implementation for an interface, we can create implementations for most things.

Without this functionality, we instead need to create a decorator/wrapper/adapter/facade which implements the interface for any values or types we want to map to an interface.

A common usage of this pattern could likely be found in ASP.NET apps.  Problem details response may be returned as a response from an action controller.   An interface could be defined with a method that returns a `ProblemDetails` object, and implementations created for validation objects and exceptions.  Likewise, we could create implementations for `IActionResult` for success objects, validation objects, and exceptions.

This has flow on effects for the testability of our applications.  Each implementation contains isolated logic which is individually testable.  An implementation is testable if the accessor modifier is sufficient to access it in a testing project.  Simply create an object of a type with an interface implementation and assign it to a variable of the interface type, then make assertions against the interface.

**The syntax**

```
<public/private/internal/file> implementation <value or ref type> : <interface> {
    // the definition of <interface> must be implemented
}
```

Here are some examples.

First, define an interface:

```c#
public interface Displayable
{
    string Display();
}
```

Then create implementations for the interface:

```c#
public implementation string : Displayable
{
    public string Display() => $"Simon say \"{this}\"";
}

Displayable str = "Hello world!";
str.Display(); // Simon says "Hello world!"
```

```c#
public implementation int : Displayable
{
    public string Display() => (this + 1).ToString();
}

Displayable i = 1;
i.Display(); // 2
```

```c#
public record Duck(string Name);

public implementation Duck : Displayable
{
    public string Display() => $"{Name} says \"Quack!\"";
}

Displayable duck = new Duck("George");
duck.Display(); // George says "Quack!"
```

```c#
public enum Day { Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday }

public implementation Day : Displayable
{
    public string Display() => this switch
        {
            Wednesday => "Hump day",
            Friday => "Almost the weekend!!!",
            Saturday => "Party!!!!!",
            Sunday => "Party!!!!!",
            _ => "Work, eat, sleep, repeat...",
        };
}

Displayable day = Day.Tuesday;
day.Display(); // "Work, eat, sleep, repeat..."
```

```c#
public implement int[] : Displayable
{
    public string Display() => string.Join(", ", this);
}

Displayable arr = new int[] { 2, 4, 6, 8 };
arr.Display(); // 2, 4, 6, 8
```

```c#
public class Generic<T>
{
    public T Thing { get; set; }
}

public implement Generic<T> : Displayable
{
    public string Display() => Thing.ToString()
}

public implement Generic<int> : Displayable
{
    public string Display() => $"The Generic value is {this}"
}

public implement Generic<Displayable> : Displayable
{
    public string Display() => Thing.Display();
}

Displayable g1 = new Generic<bool> { Thing = true }
g1.Display(); // True

Displayable g2 = new Generic<int> { Thing = 1 };
g2.Display(); // The Generic value is 1

Displayable g3 = new Generic<Day> { Thing = Day.Saturday };
g3.Display(); // Party!!!!!
```

```c#
public Displayable GetDisplayable() => new Generic<Day> { Thing = Day.Wednesday };

var x = GetDisplayable();
x.Display(); // Hump day
```

```c#
public Displayable GetDisplayable<T>(T val) => new Generic<T> { Thing = val  };

var x = GetDisplayable(1);
x.Display(); // The Generic value is 1
```

```c#
public implementation Exception : Displayable
{
    public string Display() => this.Message;
}

public Displayable TryDivideOne(int val) {
    try { return 1 / val; }
    catch Exception e { return e; }
}

var val = TryDivideOne(0);
val.Display(); // Division by constant zero
```

**Comment by bernd5** [comment]
This is maybe part of https://github.com/dotnet/csharplang/blob/main/proposals/extensions.md

**Reply by br3nt** [reply]
There is a difference in that extensions extend and add to the members that can be accessed for a value or type. 

This doesn't do that. Instead, to access the implementation, the value or type must be assigned or cast to the interface.

I've updated the description in the hopes to make the differences more clear.

**Comment by CyrusNajmabadi** [comment]
This feels just like extensions: https://github.com/dotnet/csharplang/blob/main/proposals/extensions.md

**Reply by br3nt** [reply]
Do extensions code to an interface?

Don't extensions just add methods or functionality to existing things?

Can you define an extension such that the extended thing can be assigned to variable of a given interface?  I think the answer is yes but I'm not sure. 

I guess the difference between extensions is that things would need to be cast or assigned to the interface.

**Reply by Joe4evr** [reply]
> Can you define an extension such that the extended thing can be assigned to variable of a given interface? I think the answer is yes but I'm not sure.

The plan is 'Yes', but not in Phase A.

**Reply by br3nt** [reply]
Ok, that was quite a long read.

I think what that is saying is that any value or type can be extended with additional methods and properties and what not. Like extension methods, but can define multiple things. The extensions get added to the value or type being extended. So create an extension for an enum, and any value of that enum gains all the members defined in the extension.

This proposal isn’t extensions. Nothing is being added to the value or type. It is an interface definition for the value or type. This allows that value or type to be passed to a parameter of the type of that interface, or assigned to a variable of the type of the interface, or cast to the type of that interface.

Using the `Displayable` interface and `Day` enum from the description, you couldn’t do `Day.Wednesday.Display()` because we aren’t extending or adding anything to the value or type. 

But you could do one of the following:

```c#
(Day.Wednesday as Displayable).Display();
```

Or

```c#
public void MyFunc(Displayable d) { }

MyFunc(Day.Wednesday);
```

Or

```c#
Displayable day = Day.Wednesday;
```

Or

```c#
Day.Wednesday is Displayable ? "Displayable" : "Not displayable"; // Displayable
```

Does that make sense?

**Reply by 333fred** [reply]
> This allows that value or type to be passed to a parameter of the type of that interface, or assigned to a variable of the type of the interface, or cast to the type of that interface.

But extensions allows that too, and more. In theory, you'll even be able to explicitly implement the interface in the extension such that you must cast the extended type to the interface to access the interface members. This proposal just seems wholely subsumed by something we're already working on.
